docs/src/middleware/broker.md

```diff
@@ -0,0 +1,72 @@
+# Middleware Broker
+
+Middleware broker acts as an app-level integration gateway in Valence Programs.
+*Integration* here is used rather ambiguously on purpose - brokers should remain
+agnostic to the primitives being integrated into Valence Protocol. These primitives
+may involve but not be limited to:
+
+- data types
+- functions
+- encoding schemes
+- any other distributed system building blocks that may be implemented differently
+
+### Problem statement
+
+Valence Programs can be configured to span over multiple domains and last for an
+indefinite duration of time.
+
+Domains integrated into Valence Protocol are sovereign and evolve on their own.
+
+Middleware brokers provide the means to live with these differences by enabling
+various primitive conversions to be as seamless as possible. Seamless here primarily
+refers to causing **no downtime** to bring a given primitive up-to-date, and
+making the process of doing so as **easy as possible** for the developers.
+
+To visualize a rather complex instance of this problem, consider the following
+situation. A Valence Program is initialized to continuously query a particular
+type from a remote domain, modify some of its values, and send the altered object
+back to the remote domain for further actions.
+At some point during the runtime, remote domain performs an upgrade which extends
+the given type with additional fields. The Valence Program is unaware of this
+upgrade and continues with its order of operations. However, the type in question
+from the perspective of the Valence Program had drifted and is no longer
+representative of its origin domain.
+
+Among other things, Middleware brokers should enable such programs to gracefully
+recover into a synchronized state that can continue operating in a correct manner.
+
+## Broker Lifecycle
+
+Brokers are singleton components that are instantiated before the program start
+time.
+
+Valence Programs refer to their brokers of choice by their respective addresses.
```

Comment:
Which component of Valence Programs should refer to the broker?

Reply:
any component is free to do that.

if it's a library, then the logic needs to be encoded in the actual code (querying the broker and performing actions based on the response).

if it's a processor, then an assumption is made that it is the one initiating the queries and evaluating the responses (in some kind of stack-resembling fashion). if the evaluation checks out, some action may be performed.